### DISCUSSION TITLE
New mod idea: Adaptive Speed

### DISCUSSION THREAD

**Discussion by hlysine** [ppy/osu#17028] [Ideas] [answered]
This mod changes track speed during gameplay according to your hit error. If you consistently hit too early, the track will speed up to catch up with you, and vice versa. Basically, you decide the BPM of the track by playing at that speed. 

Like other rate-changing mods, you can choose whether the speed change affects pitch as well. You may also choose the initial rate (from 0.5x to 2x).

You can use this mod as:
- a challenge for yourself to keep good accuracy and steady rhythm throughout a song (since the game no longer maintains the BPM for you)
- a practice mode to see how fast you can go/how slow you need to go to pass a map

I have got a rough implementation of this mod already, mostly following how `ModTimeRamp` is done. Just opening this discussion for opinions/suggestions before I continue working on it.

**Comment by apollo-dw** [comment]
this is going to be hell to account for on the difficulty calculation side of things lol

**Reply by Axilotl17** [reply]
not if they put it in the fun, unranked category like wind up

**Reply by bdach** [reply]
Unranked mods no longer exist in lazer.

**Reply by Sublimelime** [reply]
> Unranked mods no longer exist in lazer.

Out of the loop here - does this mean that in the future when lazer awards pp, all the mods that exist will adjust pp accordingly? So even silly or "for fun"-type ones will just have a multiplier on pp? If so that's pretty cool!

**Reply by GopoBoi** [reply]
no, the only thing that really means is that all plays will be sent to the server. all mods will probably have a leaderboard, but whether a mod will give pp is up to the score processor to determine.

**Reply by abraker95** [reply]
> this is going to be hell to account for on the difficulty calculation side of things lol

This mod adjusts the map rate so that in theory difficulty due to speed converges to 0. So the two main ?? are:
* How fast does it converge in response to change in rate the player can't do? If it's very responsive, there is a possibility of a feedback loop that causes oscillation in rate change. If it's slow, then there should be an interval following note speed change where difficulty due to speed is still relevant.
* Do the changes cause reading and rhythmic complexity challenges?

That said, first question is relevant from gameplay point-of-view. Can this mod actually cause the rate to oscillate as the player tries to fight it to hit the notes on time? I can imagine that might be quite annoying gameplay wise.

**Reply by bdach** [reply]
> Can this mod actually cause the rate to oscillate as the player tries to fight it to hit the notes on time? I can imagine that might be quite annoying gameplay wise.

it can yes. that's sort of the point.

**Reply by abraker95** [reply]
Is it really? I'd imagine oscillation is an undesired side effect.

**Reply by bdach** [reply]
as implemented it is because the rate is calculated using a sliding window, which prevents convergence to a single final rate. @hlysine will be able to comment better on why this was done that way but i kind of like it, otherwise this would become just a normal rate mod late on. it's not supposed to be an overly serious mod anyway.

**Reply by abraker95** [reply]
Not saying that a sliding window shouldn't be used, but rather a good dampening value be chosen that balances between being too responsive and converging to one value for too long. That or let dampening be a setting.

**Reply by hlysine** [reply]
This mod is designed to magnify your mistakes in accuracy and make you work towards fixing those mistakes, so if you play a map with bad accuracy, oscillation is expected and you have to fight it. To prevent the track rate from being overly responsive, the actual track rate is calculated by averaging the rate of the last 6 hits, so occasional inaccuracies won't completely mess up the rate.

**Comment by frenzibyte** [comment, marked as answer]
Implementation PR'd at #17054.
